lms/djangoapps/courseware/index.py

```diff
@@ -0,0 +1,554 @@
+    """
+    View class for the Courseware page.
+    """
+    @method_decorator(login_required)
+    @method_decorator(ensure_csrf_cookie)
+    @method_decorator(cache_control(no_cache=True, no_store=True, must_revalidate=True))
+    @method_decorator(ensure_valid_course_key)
+    def get(self, request, course_id, chapter=None, section=None, position=None):
+        """
+        Displays courseware accordion and associated content.  If course, chapter,
+        and section are all specified, renders the page, or returns an error if they
+        are invalid.
+
+        If section is not specified, displays the accordion opened to the right
+        chapter.
+
+        If neither chapter or section are specified, displays the user's most
+        recent chapter, or the first chapter if this is the user's first visit.
+
+        Arguments:
+            request: HTTP request
+            course_id (unicode): course id
+            chapter (unicode): chapter url_name
+            section (unicode): section url_name
+            position (unicode): position in module, eg of <sequential> module
+        """
+        self.course_key = CourseKey.from_string(course_id)
+        self.request = request
+        self.original_chapter_url_name = chapter
+        self.original_section_url_name = section
+        self.chapter_url_name = chapter
+        self.section_url_name = section
+        self.position = position
+        self.chapter, self.section = None, None
+
+        try:
+            self._init_new_relic()
+            self._verify_position()
+            with modulestore().bulk_operations(self.course_key):
+                self.course = get_course_with_access(request.user, 'load', self.course_key, depth=CONTENT_DEPTH)
+                self.is_staff = has_access(request.user, 'staff', self.course)
+                self._setup_masquerade_for_effective_user()
+                return self._get()
+        except Redirect as redirect_error:
+            return redirect(redirect_error.url)
+        except UnicodeEncodeError:
+            raise Http404("URL contains Unicode characters")
+        except Http404:
+            # let it propagate
+            raise
+        except Exception:  # pylint: disable=broad-except
+            return self._handle_unexpected_error()
+
+    def _setup_masquerade_for_effective_user(self):
+        """
+        Setup the masquerade information to allow the request to
+        be processed for the requested effective user.
+        """
+        self.real_user = self.request.user
+        self.masquerade, self.effective_user = setup_masquerade(
+            self.request,
+            self.course_key,
+            self.is_staff,
+            reset_masquerade_data=True
+        )
+        # Set the user in the request to the effective user.
+        self.request.user = self.effective_user
+
+    def _get(self):
+        """
+        Render the index page.
+        """
+        self._verify_course_entry()
```

Comment:
Fair enough.  But `course_access` has different meaning in the platform, referring to the checks made in access.py (like start-date, visibility, is_staff, etc). 

And since this method, in fact, redirects the user to fix the problem, how about something like `redirect_to_gain_course_entry_if_needed'?